### PR TITLE
Bugfix/flash clash

### PR DESCRIPTION
**Pull request by 0beron**
Validation on read of the bytes that were moved. Should be low risk for users upgrading to 1.0.1 and staying there, but this can protect against problems that result from loading a pre 1.0.1 firmware and reading arbitrary data into some default settings.

### CODE REVIEW

**Review by m-m-adams — APPROVED** [review]
Looks good and works on my deluge, I'd like another check before merging for sanity though

**Comment by m-m-adams** [comment]
This conflicts with the other fix, I think you need to rebase

File: src/deluge/storage/flash_storage.cpp [hunk 1]
```diff
@@ -420,6 +432,44 @@ void readSettings() {
...
+	midiEngine.midiFollowFeedback = buffer[129];
+	midiEngine.midiFollowFeedbackAutomation = static_cast<MIDIFollowFeedbackAutomationMode>(buffer[130]);
+	midiEngine.midiFollowFeedback = buffer[131];
+	*/
+
+	/* Bytes 132-134 have migrated from a previous location. Loading firmware either side
+	   of the change sequentially can risk reading arbitrary data, which will
+	   cast to an enum of the right type, but with a value that matches none of
+	   the available enum values. Pick a valid default if this happens. */
+	defaultSessionLayout = static_cast<SessionLayoutType>(buffer[132]);
```
Comment (bobtwinkles) [inline src/deluge/storage/flash_storage.cpp:454]: I think we should test these with a pattern like

```cpp
if (util::to_underlying(SessionLayoutTypeMaxElement) <= buffer[132]) {
  defaultSessionLayout = SessionLayoutType::SessionLayoutTypeRows;
} else {
  defaultSessionLayout = static_cast<SessionLayoutType>(buffer[132]);
}
```

This way we avoid ever having bit patterns that represent invalid enum values in a variable with that enum's type.

Comment (0beron) [reply src/deluge/storage/flash_storage.cpp:454]: Shall I apply this to the other enum class based flash settings that go further back? These have all worked OK in the past, just defaultSessionLayout and defaultKeyboardLayout are vulnerable, especially if you switch between firmwares that straddle the branch to 1.0.1

**Comment by seangoodvibes** [comment]
hi @0beron could you please update the placeholders for midi follow and expand them to match the following? Thanks

> 123: midiFollow enable/disable
> 124: midiFollow set follow channel synth
> 125: midiFollow set follow channel kit
> 126: midiFollow set follow channel param
> 127: midiFollow set kit root note
> 128: midiFollow display param pop up
> 129: midiFollow feedback
> 130: midiFollow feedback automation mode
> 131: midiFollow feedback filter to handle feedback loops
> 132-135: midiFollow set follow device synth 	product / vendor ids
> 136-139: midiFollow set follow device kit		product / vendor ids
> 140-143: midiFollow set follow device param		product / vendor ids

Read:

> midiEngine.midiFollow = buffer[123];
> 	midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::SYNTH)].channelOrZone = buffer[124];
> 	midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::KIT)].channelOrZone = buffer[125];
> 	midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::PARAM)].channelOrZone = buffer[126];
> 	midiEngine.midiFollowKitRootNote = buffer[127];
> 	midiEngine.midiFollowDisplayParam = buffer[128];
> 	midiEngine.midiFollowFeedback = buffer[129];
> 	midiEngine.midiFollowFeedbackAutomation = static_cast<MIDIFollowFeedbackAutomationMode>(buffer[130]);
> 	midiEngine.midiFollowFeedbackFilter = buffer[131];
> 	MIDIDeviceManager::readMidiFollowDeviceReferenceFromFlash(MIDIFollowChannelType::SYNTH, &buffer[132]);
> 	MIDIDeviceManager::readMidiFollowDeviceReferenceFromFlash(MIDIFollowChannelType::KIT, &buffer[136]);
> 	MIDIDeviceManager::readMidiFollowDeviceReferenceFromFlash(MIDIFollowChannelType::PARAM, &buffer[140]);

Write:

> 	buffer[123] = midiEngine.midiFollow;
> 	buffer[124] = midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::SYNTH)].channelOrZone;
> 	buffer[125] = midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::KIT)].channelOrZone;
> 	buffer[126] = midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::PARAM)].channelOrZone;
> 	buffer[127] = midiEngine.midiFollowKitRootNote;
> 	buffer[128] = midiEngine.midiFollowDisplayParam;
> 	buffer[129] = midiEngine.midiFollowFeedback;
> 	buffer[130] = util::to_underlying(midiEngine.midiFollowFeedbackAutomation);
> 	buffer[131] = midiEngine.midiFollowFeedbackFilter;
> 	MIDIDeviceManager::writeMidiFollowDeviceReferenceToFlash(MIDIFollowChannelType::SYNTH, &buffer[132]);
> 	MIDIDeviceManager::writeMidiFollowDeviceReferenceToFlash(MIDIFollowChannelType::KIT, &buffer[136]);
> 	MIDIDeviceManager::writeMidiFollowDeviceReferenceToFlash(MIDIFollowChannelType::PARAM, &buffer[140]);

File: src/deluge/storage/flash_storage.cpp (hunk 1 above)
Comment (bobtwinkles) [reply src/deluge/storage/flash_storage.cpp:454]: Up to you. My preference is slightly towards fixing it for all cases, but we can likely trust it won't misbehave in practice for anyone but developers for those but it's a potential problem vector.

**Comment by 0beron** [comment]
Closing this PR in favour of #868.
Trying to reserve stuff for unmerged future PRs deemed futile.

**Pull request closed without merging**